tests/integration/conftest.py

```diff
@@ -0,0 +1,200 @@
+    """Get the jenkins agent image."""
+    value: None | str = pytestconfig.getoption("--jenkins-agent-image")
+    assert value is not None, "please specify the --jenkins-agent-image command line option"
+    return value
+
+
+@pytest.fixture(scope="module")
+def jenkins_controller_name(pytestconfig: pytest.Config):
+    """Get the name of the controller jenkins is running on."""
+    value: None | str = pytestconfig.getoption("--jenkins-controller-name")
+    assert value is not None, "please specify the --jenkins-controller-name command line option"
+    return value
+
+
+@pytest.fixture(scope="module")
+def jenkins_model_name(pytestconfig: pytest.Config):
+    """Get the name of the model jenkins is running on."""
+    value: None | str = pytestconfig.getoption("--jenkins-model-name")
+    assert value is not None, "please specify the --jenkins-model-name command line option"
+    return value
+
+
+@pytest.fixture(scope="module")
+def jenkins_unit_number(pytestconfig: pytest.Config):
+    """Get the number of the unit jenkins is running on."""
+    value: None | str = pytestconfig.getoption("--jenkins-unit-number")
+    assert value is not None, "please specify the --jenkins-unit-number command line option"
+    assert value.isdigit(), "--jenkins-unit-number must be a non-negative integer"
+    return int(value)
+
+
+@pytest.fixture(scope="module")
+def metadata():
+    """Provides charm metadata."""
+    yield yaml.safe_load(pathlib.Path("./metadata.yaml").read_text(encoding="utf-8"))
+
+
+@pytest.fixture(scope="module")
+def app_name(metadata):
+    """Provides app name from the metadata."""
+    yield metadata["name"]
+
+
+@pytest_asyncio.fixture(scope="module")
+async def agent_label(ops_test: OpsTest):
+    """Provides label that uniquely identifies the agent under test."""
+    yield ops_test.model_name
+
+
+@pytest_asyncio.fixture(scope="module")
+async def app(
+    ops_test: OpsTest,
+    app_name: str,
+    jenkins_agent_image: str,
+    jenkins_controller_name: str,
+    jenkins_model_name: str,
+    agent_label: str,
+):
+    """Charm used for integration testing.
+
+    Builds the charm and deploys it and the relations it depends on.
+    """
+    # This helps with type hints, it seems model could be None
+    assert ops_test.model is not None
+
+    # Build and deploy the charm
+    charm = await ops_test.build_charm(".")
+    resources = {"jenkins-agent-image": jenkins_agent_image}
+    application = await ops_test.model.deploy(
+        charm,
+        resources=resources,
+        application_name=app_name,
+        series="jammy",
+        config={"jenkins_agent_labels": agent_label},
+    )
+    await ops_test.model.wait_for_idle()
+
+    # Create the relationship
+    jenkins_controller_model_name = f"{jenkins_controller_name}:{jenkins_model_name}"
+    await ops_test.model.create_offer(application_name=app_name, endpoint=f"{app_name}:slave")
+    subprocess.check_output(
```

Comment:
Instead of running os calls, it'd be better to use the pytest_operator API. You can see some examples in [the Indico charm](https://github.com/canonical/indico-operator/tree/main/tests/integration)

Reply:
Normally that is how it should be done, in this case the problem is that the relation is not only cross-model, but cross-controller (since Jenkins runs on lxd and the Jenkins Agent on Kubernetes). When trying to run pytest-operator on a controller that runs on lxd and also controls a kubernetes deployment, pytest operator crashes reporting kubernetes related problems. It is possible to run jenkins agent outside of pytest operator so the problem is specifically related to pytest operator.

Reply:
I haven't found any APIs in pytest-operator that allow cross controller relations to be created. Happy to use them if there is one

Reply:
Thank you to @weiiwang01, instead of using `subprocess` the calls to juju now use `ops_test.juju` where there isn't native support for taking a given action (e.g., for the cross-controller operations required for Jenkins interactions).